### PR TITLE
Wps results fix

### LINKED ISSUE #6392: WPS results are showing columns twice after sharing
https://beta.geo-rapp.terria.io/#share=s-eLoi6D4gSxt6RWMxWgRlxJD0D0s

image

### PR DESCRIPTION
**Pull request by sixlighthouses**
### Fix WPS duplicate 

Fixes #6392

Workbench items were being duplicated when loaded from WPS results in share URL.  
Added a check in models/Terria.ts when loading items to ensure duplicates are not added

### Test me

Load this share URL http://ci.terria.io/wps-results-fix/#share=s-uCLMJuXtXBpFXvmJbBI4gE4rGUz

Two items should be loaded to the workbench 
image

### Checklist

- [ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
- [x] I've updated relevant documentation in `doc/`.
- [x] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

**Comment by na9da** [comment]
Hey @sixlighthouses, that's some good sleuthing :detective: job.

The root cause is that our workbench implementation is not protected from duplicate items - that might need a rethink at some point.

The fix you made works, but  I would suggest changing [newItems](https://github.com/TerriaJS/terriajs/blob/e46799735edc134acf49856995e2f263ea0556f3/lib/Models/Terria.ts#L1799) to a [Set&lt;BaseModel&gt;](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead? It'll automatically guard against similar bugs in other code paths. You can then convert the set to array using `[...newItems]`.

**Comment by sixlighthouses** [comment]
thanks @na9da a set is a nice solution will make that change

**Comment by na9da** [comment]
Hey @sixlighthouses, on another thought the better fix would be to make the set method for [workbench.items](https://github.com/TerriaJS/terriajs/blob/e46799735edc134acf49856995e2f263ea0556f3/lib/Models/Workbench.ts#L36-L38)  sanitize its input by de-duplicating the input array. You could use the lodash `uniq` method to do that.

Sorry if you already started on the fix.

**Comment by sixlighthouses** [comment]
thanks @na9da updated the workbenches items set method

File: lib/Models/Terria.ts [hunk 1]
```diff
@@ -1802,7 +1802,10 @@ export default class Terria {
     while (true) {
       const model = newItemsRaw.shift();
       if (model) {
-        await this.pushAndLoadMapItems(model, newItems, errors);
+        // check if model is already in the newItems list by uniqueId
+        if (!newItems.find((item) => item.uniqueId === model.uniqueId)) {
+          await this.pushAndLoadMapItems(model, newItems, errors);
+        }
```
Comment (na9da) [inline lib/Models/Terria.ts]: I guess we could remove this check now that our workbench will weed out the duplicates.

File: lib/Models/Workbench.ts [hunk 2]
```diff
@@ -34,7 +34,12 @@ export default class Workbench {
     return this._items.map(dereferenceModel);
   }
   set items(items: readonly BaseModel[]) {
-    this._items.spliceWithArray(0, this._items.length, items.slice());
+    const setItems = new Set(items);
+    this._items.spliceWithArray(
```
Comment (na9da) [inline lib/Models/Workbench.ts:39]: Probably add a comment here to say that we want the workbench items to be unique.

**Comment by na9da** [comment]
Looks good @sixlighthouses. I have just made a couple of comments.

File: CHANGES.md [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
-#### next release (8.3.2)
+#### next release (8.3.3)
 
 - Fixed a bug when restoring timefilter from a share link having more than one imagery item with the same base URL (but different layer names).
 - [The next improvement]
 
+#### 8.3.2
+
+- Fix WPS duplicate display of analysis results when loaded through a share URL
+
```
Comment (na9da) [inline CHANGES.md]: I think the correct next release for this change is `8.3.2` (so after the timefilter bug fix).

https://github.com/TerriaJS/terriajs/blob/15d468656bee43b0f872075cbf867475993e2536/CHANGES.md?plain=1#L3-L7

File: lib/Models/Terria.ts (hunk 1 above)
Comment (sixlighthouses) [reply lib/Models/Terria.ts]: reverted back to previous method

File: lib/Models/Workbench.ts (hunk 2 above)
Comment (sixlighthouses) [reply lib/Models/Workbench.ts:39]: good idea, commented

**Review by na9da — APPROVED** [review]
Good to go :rocket:

**Pull request merged**